go/mcap/parse.go

```diff
@@ -264,41 +264,40 @@ func ParseChunkIndex(buf []byte) (*ChunkIndex, error) {
 	}, nil
 }
 
-// ParseAttachment parses an attachment record.
-func ParseAttachment(buf []byte) (*Attachment, error) {
-	logTime, offset, err := getUint64(buf, 0)
+func parseAttachmentReader(r io.Reader) (*AttachmentReader, error) {
+	buf := make([]byte, 8)
+	crcReader := newCRCReader(r)
+	logTime, err := readUint64(buf, crcReader)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read record time: %w", err)
 	}
-	createTime, offset, err := getUint64(buf, offset)
+	createTime, err := readUint64(buf, crcReader)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read create time: %w", err)
 	}
-	name, offset, err := readPrefixedString(buf, offset)
+	name, err := readPrefixedString(buf, crcReader)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read attachment name: %w", err)
 	}
-	mediaType, offset, err := readPrefixedString(buf, offset)
+	mediaType, err := readPrefixedString(buf, crcReader)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read media type: %w", err)
 	}
-	dataSize, offset, err := getUint64(buf, offset)
+	dataSize, err := readUint64(buf, crcReader)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read attachment data size: %w", err)
 	}
-	data := buf[offset : offset+int(dataSize)]
-	offset += int(dataSize)
-	crc, _, err := getUint32(buf, offset)
-	if err != nil {
-		return nil, fmt.Errorf("failed to read CRC: %w", err)
-	}
-	return &Attachment{
-		LogTime:    logTime,
-		CreateTime: createTime,
-		Name:       name,
-		MediaType:  mediaType,
-		Data:       data,
-		CRC:        crc,
+	limitReader := io.LimitReader(crcReader, int64(dataSize))
```

Comment:
how do you feel about implementing a custom `LimitedReadSeeker` and returning that instead? The LimitReader source is not particularly complicated, it doesn't feel too hard to adapt.

Reply:
what's the need for seeking?

Reply:
also, I think that's going to require retention of the already-read data in memory if it doesn't exist on disk. Exposing io.Reader seems correct to me here; the limit aspect is just to ensure it EOFs at the right place.

Reply:
If the underlying reader is a Seeker (which we already specialize for in some cases), then it may make sense to hand the user the ability to seek through the attachment for parts that they want.

Reply:
I'm struggling to see a way to make this work that won't seem clunky or work against the CRC validation features.

* CRC validation seems like it would definitely break as soon as the user seeks around the attachment. If they don't read the whole thing, or read it in the wrong order, their CRC isn't going to validate.
* I don't think AttachmentReader should type the Data section with an io.ReadSeeker if there is any possibility that the thing cannot Seek (which is the case when doing a streaming read). So I think it would return an io.Reader, and if we got this working in some manner it would be up to the caller to attempt to cast to a ReadSeeker, and if successful seek away and break their own CRC validation.
* The brute force mechanism I think would be to make CRCReader a public type, and expose its underlying reader. Since the underlying reader is a LimitedReader that also exposes its underlying reader, by just making the CRCReader public the user would be able to call, effectively, `rs, ok := ar.Data().(CRCReader).R.(LimitedReader).R.(io.ReadSeeker)` and seek on that if successful. That would actually need to be split across multiple type assertions in case one of the intermediate ones fail. Definitely seems too clunky to me to ship.

Open to ideas on clean ways to handle this. I don't think it's a solution to just implement a LimitedReadSeeker and return that (even behind the io.Reader) because in common cases it would not actually be seekable. My inclination at the moment is to keep this as is and force the reader to consume the whole attachment (assuming they want any of it), and wait for feedback before optimizing. I think the usual case for dealing with attachments in mcap files is going to be to dump them out to disk for post-processing rather than do seeking reads on them in place - but we won't know till it gets some use.

If users have a Seekable reader they also have the option of addressing the attachment via the index and seeking around however they want, which the SDK fully supports. IMO that is probably enough to cover this usecase for seeking reads on an attachment in-place.